### DISCUSSION TITLE
Need help in adding EXIF data to original photos

### DISCUSSION THREAD

**Discussion by gspannu** [RhetTbull/osxphotos#1250] [Q&A]
My setup:
- MacBook Pro (Apple Silicon)
- Photos using Referenced files (residing on same hard disk)
- No iCloud files, all files are local on hard disk
- Files are mainly RAW files or iPhone (HEIC) files
- Installed osxphotos as binary file
- Installed exiftool through homebrew

Workflow:
- Import files to Photos to a specific album
- keyword added
- Photo editing done as required

**Requirements (What I need some help for)**
For all photos just added to this newly created album
1. Add information from macOS Photos to _EXIF metadata in the original referenced files_.
2. Write to original files, no backup files needed, no additional sidecar files to be created
3. Run as a two stage operation

------------------------------------
**Step 1 - First pass...**
**Write EXIF data to be written to original files...**
- Add location information to Photos EXIF data (i.e. Location, City, County, Country, CountryCode)
- Add keywords to EXIF
- Add specific hardcoded values to certain EXIF fields
   Creator: "John Smith"
   Copyright: "Copyright © John Smith"
   Rights Usage: "All Rights Reserved"
- Add Title to EXIF {dynamically created - should be equal to "Location, City, Country" }

Ensure that this newly written EXIF data is now read back by Photos

**Step 2 - I will now manually edit the Titles in macOS Photos(if required and make it more suitable)**

**Step 3 - Second Pass...**
**Write EXIF data to be written to original files...**
- Caption field: {Same as Title}
- Subject field: {Same as Title}
- Headline field: {Same as Title}

Ensure that this newly written EXIF data is now read back by Photos

------------------------------------

I am trying to read the documentation and understand the various command line options, but would **appreciate if anyone could provide me the command line options for Step 1 and Step 3**.

Thanks...

**Comment by RhetTbull** [comment]
Hi. This is not a use case supported by osxphotos. osxphotos does have a [osxphotos push-exif](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-push-exif) command that "pushes" (writes) metadata from photos to the original file in the library. This would partially work for step 1 and step 3. I say partially because this command does give some control over the metadata written it does not deal with creator, copyright, etc. You can't specify custom fields (but you can specify custom values for the supported fields).  The options, available in the [docs](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-push-exif) are:

- all: all metadata
- keywords: keywords/tags (e.g. IPTC:Keywords, etc.)
- location: location information (e.g. EXIF:GPSLatitude, EXIF:GPSLongitude, etc.)
- faces: face region information (e.g. XMP:RegionName, etc.)
- persons: person in image information (e.g. XMP:PersonInImage, etc.)
- datetime: date/time information (e.g. EXIF:DateTimeOriginal, etc.)
- title: title information (e.g. XMP:Title, etc.)
- description: description information (e.g. XMP:Description, etc.)

There is currently no option to read the EXIF data back into Photos (with exception of date/time/timezone which Photos can sometimes mess up -- see [osxphotos timewarp](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-timewarp) which has a `--pull-exif` option to "pull" or read the EXIF data from the file back into Photos.  Why don't you do all these EXIF edits **before** importing the photos to Photos? Photos will read on import keyword, title, description, location, and the normal EXIF lens/shot data.

There's also a [batch-edit](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-batch-edit) command that allows you to batch edit title, keywords, etc. (and has a template system to dynamically set these) but it does not support some of the fields you want to use. 

If you know how to write code in python, you could create a scrip to do this exact workflow using the components provided by osxphotos. For example the [ExifTool](https://rhettbull.github.io/osxphotos/API_README.html#exiftoolExifTool) class provides a way to use exiftool to read & write metadata.

exiftool can do everything you want if you learn how to use it. Here's a way to get a list of files you just imported into a specific album for example:

`osxphotos query --added-in-last "2 hours" --album "YourAlbum" --quiet --print "{photo.path}"`

Obviously you change the time frame from "2 hours" to whatever makes sense for you and the album name from "YourAlbum" to the album you're working in. This prints a list of all the photo paths to the original files which can then be fed to exiftool to write additional fields. Keep in mind that this is the path to the original referenced photo. When Photos edits a photo, it keeps the edited version in a different path inside the library, not with the original referenced image.

It sounds like you're trying to use Photos as a digital asset management tool it isn't designed to be.  You might be better off with Lightroom or DigiKam, etc. 

For reference, these are the EXIF/IPTC/XMP fields that osxphotos writes to which as you can see does not include all the fields you want:

            IPTC:CaptionAbstract
            XMP:Title
            IPTC:ObjectName
            XMP:TagsList
            IPTC:Keywords (may include album name, person name, or template)
            XMP:Subject (set to keywords + person)
            XMP:PersonInImage
            EXIF:GPSLatitudeRef, EXIF:GPSLongitudeRef
            EXIF:GPSLatitude, EXIF:GPSLongitude
            EXIF:GPSPosition
            EXIF:DateTimeOriginal
            EXIF:OffsetTimeOriginal
            EXIF:ModifyDate
            IPTC:DigitalCreationDate
            IPTC:DateCreated
            QuickTime:CreationDate
            QuickTime:CreateDate (UTC)
            QuickTime:ModifyDate (UTC)
            QuickTime:GPSCoordinates
            UserData:GPSCoordinates

**Reply by RhetTbull** [reply]
By the way, there is an open issue (see #243) to add an `--exiftool-template` option to the `push-exif` and `export` commands which would allow you to specify arbitrary metadata fields to write to the file (With templated values). This could probably solve your use case once implemented.  Not sure when I'll get to implementing this.

**Reply by oPromessa** [reply]
Hey @RhetTbull. Just to be documented.

### EXIF:DateTimeOriginal (for Photos) and QuickTime:CreationDate (for Videos):
Normally:
* Photos: JPEG/HEIC have EXIF:DateTimeOriginal (and sometimes EXIF:OffsetTimeOriginal)
* Videos: MOV/MP4 have QuickTime:CreationDate 
   * and inclusively you're not able to write EXIF:DateTimeOriginal  into Videos.

* But I found/have some Canon MOV files (Canon PowerShot SX40 HS) named MVI_NNNN.MOV which:
   * have both EXIF:DateTimeOriginal and QuickTime:CreationDate!!!
   * but no EXIF:OffsetTimeOriginal

### osxphotos push-exif and timewarp -c 
* As you mentioned above (and the code below), priority is given to EXIF:DateTimeOriginal will cause `osxphotos timewarp -c` to indicate wrong TZ/UTC status between Photos/File EXIF data. Even though `osxphotos push-exif datetime` does the push correctly ;)
https://github.com/RhetTbull/osxphotos/blob/50e22ea64760e5338f08f6325779536a1ec277e0/osxphotos/exifutils.py#L55-L76

### Workaround
* I've addressed this by force setting all dates with exiftool on the actual file as per below, after I've done `push-exif date time,location`
   * [exif.MyDateKeysCreationDate.pl.REMOVE.txt.extension.txt](https://github.com/user-attachments/files/16923275/exif.MyDateKeysCreationDate.pl.REMOVE.txt.extension.txt)
   * And then I go all the way (maybe to the extreme I know) and set all dates:
```py
exiftool -config exif.MyDateKeysCreationDate.pl -overwrite_original_in_place -api QuickTimeUTC=1 -m -v0 "-DateTimeOriginal<MyDate" "-CreateDate<MyDate" "-ModifyDate<MyDate" "-FileModifyDate<MyDate" "-Track*Date<MyDate" "-Media*Date<MyDate" "-Keys:CreationDate<MyDate" "-quicktime:createdate<MyDate" "-quicktime:modifydate<MyDate" "-FileCreateDate<MyDate" "-Quicktime:ContentCreateDate<MyDate" -if 'defined $GPS:GPSDateStamp' "-GPS:GPSDateStamp<Mydate" -if 'defined $GPS:GPSTimeStamp' "-GPS:GPSTimeStamp<Mydate" -if 'defined $XMP-photoshop:DateCreated' "-XMP-photoshop:DateCreated<MyDate" -if 'defined $IPTC:DigitalCreationDate' "-IPTC:DigitalCreationDate<MyDate" -if 'defined $IPTC:DigitalCreationTime' "-IPTC:DigitalCreationTime<MyDate" -if 'defined $IPTC:TimeCreated' "-IPTC:DateCreated<MyDate" -if 'defined $IPTC:TimeCreated' "-IPTC:TimeCreated<MyDate" "-AllDates<MyDate" MVI_NNNN.MOV
```
* But still not able to set EXIF:SubSecDateTimeOriginal so `timewarp -c` will still not find TZ info  on EXIF file.

**Comment by gspannu** [comment]
@RhetTbull : Thank you for your detailed response. It has given me some way forward.

I could have used Lightroom but a) it is an expensive product, b) it does not do Reverse-Geocoding.

**Are you aware of any tools (GitHub, OpenSource) that could take a photo file as input and reverse-geoencode the GPS location information contained in the photo to location, city, country, etc...**?

Essentially, what I need is the Location. City, County, Country returned as variables (or in some shape & form) - 

I am confident that I could do the rest in ExifTool.... and then as you suggest, import the photos after this processing.

I am conversant with Exiftool and use it quite often, but I am not great with python.

**Reply by RhetTbull** [reply]
> Are you aware of any tools (GitHub, OpenSource) that could take a photo file as input and reverse-geoencode the GPS location information contained in the photo to location, city, country, etc...?

I am not. So I am building one :-)

There are a number of reverse geocoding APIs but all require an API key and some require paid access. The Mac has excellent built in Location Services which any app can use.  Unfortunately these do not work from the command line (due to App sandboxing and entitlements) so I've built a prototype "location server". It runs as a menu bar app that you can use directly to plug in a longitude and latitude and get the reverse geo information but it also provides an API that can be called from your own scripts. For example, the following gets the latitude and longitude from exiftool for a file then queries the API to get the detailed reverse geo information:

```bash
exiftool -c '%.6f' -GPSPosition  ~/Desktop/IMG_4908.HEIC -json
[{
  "SourceFile": "/Users/user/Desktop/IMG_4908.HEIC",
  "GPSPosition": "33.953636 N, 118.338950 W"
}]
curl -X PUT -H "Content-Type: application/json" -d '{"latitude":33.953636, "longitude":-118.338950}' http://localhost:8000/reverse_geocode | jq .
```

produces:

```json
{
  "location": [
    33.953636,
    -118.33895
  ],
  "name": "SoFi Stadium",
  "thoroughfare": "Stadium Dr",
  "subThoroughfare": "1001",
  "locality": "Inglewood",
  "subLocality": "Century",
  "administrativeArea": "CA",
  "subAdministrativeArea": "Los Angeles County",
  "postalCode": "90305",
  "ISOcountryCode": "US",
  "country": "United States",
  "postalAddress": "<CNPostalAddress: 0x60000074edf0: street=1001 Stadium Dr, subLocality=Century, city=Inglewood, subAdministrativeArea=Los Angeles County, state=CA, postalCode=90305, country=United States, countryCode=US>",
  "inlandWater": "None",
  "ocean": "None",
  "areasOfInterest": "(\n    \"SoFi Stadium\"\n)",
  "timeZoneName": "America/Los_Angeles",
  "timeZoneAbbreviation": "PDT",
  "timeZoneSecondsFromGMT": -25200
}
```

From there we can extract the details you seek and add them to the file via exiftool. This is just a proof of concept at this point -- I intend to create a command line tool that will call the API server to get the data then automatically write to exiftool. I will also plan to add a Finder Service so that you can right click on a file and select "Add reverse geocode data" to do this in the Finder.

None of the code is posted yet -- I'll let you know when I send it to Github and it's ready to try.

Edit: an advantage of this technique is that the reverse geocode data will be exactly the same as if it came from Photos because it's using the same Location Services data used by Photos.

**Reply by gspannu** [reply]
Oh man…. You are a rock star !

Th finder service is a real gem… 

I have a few finder services to do some of my exiftool repeat tasks by executing the exiftool command in a script inside a service. And since exiftool supports multiple files as input, the execution is quick, as the underlying script executing the exiftool command is only called once, rather than per file. 
I find it quite surprising that more people haven’t written/ published Exiftool based Finder services.

Anyway, I shall happily try out the code as soon as it is written… both the command line tool that can write directly the reverse-geoencoded fields to the photo file… as well the Finder service when its ready.

Many thanks…

**Reply by RhetTbull** [reply]
What are the tag names you want to write?  I see the following as candidates:

IPTC:

- City
- Sub-location
- Province-State
- Country-PrimaryLocationCode
- Country-PrimaryLocationName

XMP: A lot of location related tags in the following groups:

- iptcCore
- iptcExt
- photoshop

**Reply by gspannu** [reply]
I normally write only to the XMP tags now (since IPTC is an older format), also the XMP tags are able to be read by lots of other software like LR, Photos, PhotoMechanic, etc...

I use the following XMP direct tags for geo-location:
-XMP:CountryCode -XMP:Country -XMP:State -XMP:City -XMP:Location -XMP:SubLocation

As an example, **once I have managed to get all my tags written to the sidecar file (*.XMP)** using various means ... _I then transfer the tags to my original RAW, or JPG files using the following exiftool command_

`exiftool -tagsfromfile %d%f.XMP -XMP:CountryCode -XMP:Country -XMP:State -XMP:City -XMP:Location -XMP:SubLocation -XMP:Rating -XMP:Subject -XMP:Creator="Guru Pannu" -XMP:UsageTerms="ALL RIGHTS RESERVED" -XMP:Rights="Copyright © Guru Pannu" -XMP:Marked=True -EXIF:Copyright="Copyright © Guru Pannu" -EXIF:Artist="Guru Pannu" "-filemodifydate<datetimeoriginal" "-modifydate<datetimeoriginal" "-metadatadate<datetimeoriginal" "-historywhen<datetimeoriginal" -overwrite_original -ext CR2 -ext JPG "$@"`

You would have notice that in the above command.. I also 
hard-code some values for XMP:Creator, XMP:UsageTerms, XMP:Rights, XMP:Marked 
and some additional hard-coded EXIF values like EXIF:Copyright, EXIF:Artist
and set all times to datetimeoriginal

I tend to use the XMP:tag directly rather than trying to write to the various XMP sub structures (_e.g. XMP LocationDetails Struct or XMP iptcCore_) ... as exiftool seems to handle this aspect really well automatically.

Having written to the XMP:tags directly (_e.g. -XMP:CountryCode -XMP:Country -XMP:State -XMP:City -XMP:Location -XMP:SubLocation_) ... I have noticed that most software (including Photos, Aperture, Preview, LightRoom) interpret this location information correctly.

**Reply by RhetTbull** [reply]
@gspannu The initial release of the new app, [Locationator](https://github.com/RhetTbull/locationator) is up [here](https://github.com/RhetTbull/locationator).  This version simply provides the reverse geocoding API server with no other tools. You can use curl to get the reverse geocoding data for a lat/lon pair and that's it.  When I have time I'll work on adding a CLI for adding data via exiftool as well as the Service. In the mean time, if you have scripting skills you should be able to cobble something together using [jq](https://github.com/jqlang/jq) to query the returned JSON results.

Example:

`curl -X PUT -H "Content-Type: application/json" -d '{"latitude":33.953636, "longitude":-118.338950}' http://localhost:8000/reverse_geocode`

```json
{
    "ISOcountryCode": "US",
    "administrativeArea": "CA",
    "areasOfInterest": [
        "SoFi Stadium"
    ],
    "country": "United States",
    "inlandWater": "None",
    "locality": "Inglewood",
    "location": [
        33.953636,
        -118.33895
    ],
    "name": "SoFi Stadium",
    "ocean": "None",
    "postalAddress": {
        "ISOCountryCode": "US",
        "city": "Inglewood",
        "country": "United States",
        "postalCode": "90305",
        "state": "CA",
        "street": "1001 Stadium Dr",
        "subAdministrativeArea": "Los Angeles County",
        "subLocality": "Century"
    },
    "postalCode": "90305",
    "subAdministrativeArea": "Los Angeles County",
    "subLocality": "Century",
    "subThoroughfare": "1001",
    "thoroughfare": "Stadium Dr",
    "timeZoneAbbreviation": "PDT",
    "timeZoneName": "America/Los_Angeles",
    "timeZoneSecondsFromGMT": -25200
}
```

**Reply by gspannu** [reply]
You are awesome... 
I shall await your CLI and the Finder service; meanwhile I will have a play with this...
👍

**Reply by RhetTbull** [reply]
I added a new GET endpoint that's easier to use than the PUT endpoint:

`curl -X GET "http://localhost:8000/reverse_geocode?latitude=33.953636&longitude=-118.338950"`

**Reply by RhetTbull** [reply]
Version [0.0.5 of Locationator](https://github.com/RhetTbull/locationator/releases/tag/v0.0.5) now has a menu option to "Install Command Line tools" that will install `locationator` CLI to `/usr/local/bin`. The CLI has 3 commands. lookup, from-exif, and write-xmp.  `from-exif` reads the GPS coordinates from a file using exiftool and prints out the reverse geocoding data. `write-xmp` reads the GPS coordinates from the file and writes the following fields:

- XMP:CountryCode (ISOcountryCode)
- XMP:Country (country)
- XMP:State
  (administrativeArea)
- XMP:City (locality)
- XMP:Location (name)

`XMP:SubLocation` does not appear to be a valid field. There is an `XMP-iptcExt:LocationCreatedSublocation`. Need more info on where to write sublocation.

```
Usage: locationator [OPTIONS] COMMAND [ARGS]...

Options:
  --debug
  --port INTEGER
  --help          Show this message and exit.

Commands:
  from-exif  Lookup the reverse geolocation for an image/video file using...
  lookup     Lookup the reverse geolocation for a lat/lon pair.
  write-xmp  Lookup the reverse geolocation for an image/video file using...
```

**Reply by gspannu** [reply]
I am at a track day car event today… testing on this tomorrow and Wed.
Many thanks for the tool in such such short time…
👍

**Reply by gspannu** [reply]
**Yes, just tested v0.0.6.... Works.** 👍 

_The sub-location filed is actually a remnant of the IPTC era_. There used to be a IPTC field called sublocation, but there is no direct sublocation tag in XMP. 

I think the current implemented (5) tags work for the purpose. Many thanks...

---
Can the CLI be extended to support multiple files (`locationator *.HEIC`) or even something like 
`locationator 01.heic 02.heic 03.jpg 05.cr2 06.heic `
